src/include/cart/api.h

```diff
@@ -407,6 +407,34 @@ crt_reply_get(crt_rpc_t *rpc)
 	return rpc->cr_output;
 }
 
+/**
+ * Return HLC timestamp of RPC was sent
+ *
+ * \param[in] req              pointer to RPC request
+ *
+ * \return                     HLC timestamp
+ */
+uint64_t
+crt_hlc_send_get(crt_rpc_t *req);
+
+/**
+ * Return HLC timestamp of RPC (new or reply) was received
+ *
+ * \param[in] req              pointer to RPC request
+ *
+ * \return                     HLC timestamp
+ */
+uint64_t
+crt_hlc_receive_get(crt_rpc_t *req);
+
+/**
+ * Return HLC timestamp of current event
+ *
+ * \return                     HLC timestamp
+ */
+uint64_t
+crt_hlc_get(void);
```

Comment:
In my opinion, the public API should simply be int crt_hlc_get(const crt_rpc_t *rpc);

And all of the send/receive should be handled internally to cart.   The public API would either grab the latest RPC time if you pass non-NULL or the latest time if you don't.

These are automatically attached to RPCs anyway, no?  The user doesn't need to know that.

Reply:
Alternatively, given we have a public `struct crt_rpc` we should just add them there and let the user read them directly, adding a function to fetch a private struct entry doesn't add anything other than overhead.

Reply:
I specially split API into several functions. For RPC we have two different timestamps. In case of rpc == NULL which timestamp I should return, send or receive?

I am not in favor of introducing two more fields into the public structure. This will increase the memory footprint and reduce the flexibility of possibility to ​​change them later.

Reply:
Regardless, it still doesn't answer my question.   I'm not seeing the value of send/receive timestamps.  I can see the value of getting last rpc timestamp.

Reply:
Disagree. I think this is very useful. If it will not be used I remove this.

Reply:
It's obvious you disagree as you haven't changed it.   But what is the use case you have in mind?  The reason we want HLC in DAOS is purely to approximate a perfectly synchronized global clock for the purpose of timestamping updates to DAOS objects.  For that, we need the latest timestamp.  I can see the utility of the latest timestamp associated with an RPC received in order to avoid excessive calls to clock_gettime but I just don't see DAOS using the timestamp of the send for anything but perhaps logging and CART can probably handle that internally on rpc send/receive.

Reply:
Doing something that's based on NTP for synchronization isn't going to be any use to DAOS is it?

The self-test code was originally intended to report on-the-wire times for RPCs, so it could be update to measure from the "send" timestamp, depending on the exact definition.

Reply:
Ok. I remove this stuff. So, all as you wish.

Reply:
I didn't mean to remove the capability to retrieve the hlc associated with a message but rather just to remove the differentiation between send/receive.   It would still be useful to avoid the clock_getttime when assigning a timestamp to an I/O by using that latest timestamp associated with an RPC.

Johann had suggested something like crt_hlc_get(crt_rpc_t *rpc); where if rpc == NULL, we get a new timestamp, otherwise, use the latest one associated with the rpc.   Though I think Ashley's suggestion to have it as a public field in the public part of the data structure may be cleaner.   We could simply keep a copy of the value in the common header that is updated on changes.